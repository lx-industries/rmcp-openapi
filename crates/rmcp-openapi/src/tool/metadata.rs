use rmcp::model::{Tool, ToolAnnotations};
use serde_json::Value;
use std::sync::Arc;

/// Internal metadata for tools generated from OpenAPI operations.
///
/// This struct contains all the information needed to execute HTTP requests
/// and is used internally by the OpenAPI server. It includes fields that are
/// not part of the MCP specification but are necessary for HTTP execution.
///
/// For MCP compliance, this struct is converted to `rmcp::model::Tool` using
/// the `From` trait implementation, which only includes MCP-compliant fields.
#[derive(Debug, Clone, serde::Serialize)]
pub struct ToolMetadata {
    /// Tool name - exposed to MCP clients
    pub name: String,
    /// Tool title - human-readable display name exposed to MCP clients
    pub title: Option<String>,
    /// Tool description - exposed to MCP clients  
    pub description: String,
    /// Input parameters schema - exposed to MCP clients as `inputSchema`
    pub parameters: Value,
    /// Output schema - exposed to MCP clients as `outputSchema`
    pub output_schema: Option<Value>,
    /// HTTP method (GET, POST, etc.) - internal only, not exposed to MCP
    pub method: String,
    /// URL path for the API endpoint - internal only, not exposed to MCP
    pub path: String,
    /// Security requirements from OpenAPI spec - internal only, not exposed to MCP
    #[serde(skip_serializing_if = "Option::is_none")]
    pub security: Option<Vec<String>>,
}

impl ToolMetadata {
    /// Check if this tool requires authentication based on OpenAPI security definitions
    pub fn requires_auth(&self) -> bool {
        self.security.as_ref().is_some_and(|s| !s.is_empty())
    }
}

/// Converts internal `ToolMetadata` to MCP-compliant `Tool`.
///
/// This implementation ensures that only MCP-compliant fields are exposed to clients.
/// Internal fields like `method` and `path` are not included in the conversion.
impl From<&ToolMetadata> for Tool {
    fn from(metadata: &ToolMetadata) -> Self {
        // Convert parameters to the expected Arc<Map> format
        let input_schema = if let Value::Object(obj) = &metadata.parameters {
            Arc::new(obj.clone())
        } else {
            Arc::new(serde_json::Map::new())
        };

        // Convert output_schema to the expected Arc<Map> format if present
        let output_schema = metadata.output_schema.as_ref().and_then(|schema| {
            if let Value::Object(obj) = schema {
                Some(Arc::new(obj.clone()))
            } else {
                None
            }
        });

        // Create annotations with title if present
        let annotations = metadata.title.as_ref().map(|title| ToolAnnotations {
            title: Some(title.clone()),
            ..Default::default()
        });

        Tool {
            name: metadata.name.clone().into(),
            description: Some(metadata.description.clone().into()),
            input_schema,
            output_schema,
            annotations,
            // TODO: Consider migration to Tool.title when rmcp supports MCP 2025-06-18 (see issue #26)
        }
    }
}
