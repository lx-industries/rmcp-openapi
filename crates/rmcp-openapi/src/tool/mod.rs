pub mod metadata;
pub mod tool_collection;

pub use metadata::ToolMetadata;
pub use tool_collection::ToolCollection;

use crate::config::Authorization;
use crate::error::Error;
use crate::http_client::HttpClient;
use crate::security::SecurityObserver;
use reqwest::header::HeaderMap;
use rmcp::model::{CallToolResult, Tool as McpTool};
use serde_json::Value;
use url::Url;

/// Self-contained tool with embedded HTTP client
#[derive(Clone)]
pub struct Tool {
    pub metadata: ToolMetadata,
    http_client: HttpClient,
}

impl Tool {
    /// Create tool with HTTP configuration
    pub fn new(
        metadata: ToolMetadata,
        base_url: Option<Url>,
        default_headers: Option<HeaderMap>,
    ) -> Result<Self, Error> {
        let mut http_client = HttpClient::new();

        if let Some(url) = base_url {
            http_client = http_client.with_base_url(url)?;
        }

        if let Some(headers) = default_headers {
            http_client = http_client.with_default_headers(headers);
        }

        Ok(Self {
            metadata,
            http_client,
        })
    }

    /// Execute tool and return MCP-compliant result
    pub async fn call(
        &self,
        arguments: &Value,
        authorization: Authorization,
    ) -> Result<CallToolResult, crate::error::ToolCallError> {
        use rmcp::model::Content;
        use serde_json::json;

        // Create security observer for logging
        let observer = SecurityObserver::new(&authorization);

        // Log the authorization decision
        let has_auth = match &authorization {
            Authorization::None => false,
            #[cfg(feature = "authorization-token-passthrough")]
            Authorization::PassthroughWarn(header) | Authorization::PassthroughSilent(header) => {
                header.is_some()
            }
        };

        observer.observe_request(&self.metadata.name, has_auth, self.metadata.requires_auth());

        // Extract authorization header if present
        let auth_header: Option<&rmcp_actix_web::transport::AuthorizationHeader> =
            match &authorization {
                Authorization::None => None,
                #[cfg(feature = "authorization-token-passthrough")]
                Authorization::PassthroughWarn(header)
                | Authorization::PassthroughSilent(header) => header.as_ref(),
            };

        // Create HTTP client with authorization if provided
        let client = if let Some(auth) = auth_header {
            self.http_client.with_authorization(&auth.0)
        } else {
            self.http_client.clone()
        };

        // Execute the HTTP request using the (potentially auth-enhanced) HTTP client
        match client.execute_tool_call(&self.metadata, arguments).await {
            Ok(response) => {
                // Check if the tool has an output schema
                let structured_content = if self.metadata.output_schema.is_some() {
                    // Try to parse the response body as JSON
                    match response.json() {
                        Ok(json_value) => {
                            // Wrap the response in our standard HTTP response structure
                            Some(json!({
                                "status": response.status_code,
                                "body": json_value
                            }))
                        }
                        Err(_) => None, // If parsing fails, fall back to text content
                    }
                } else {
                    None
                };

                // For structured content, serialize to JSON for backwards compatibility
                let content = if let Some(ref structured) = structured_content {
                    // MCP Specification: https://modelcontextprotocol.io/specification/2025-06-18/server/tools#structured-content
                    // "For backwards compatibility, a tool that returns structured content SHOULD also
                    // return the serialized JSON in a TextContent block."
                    match serde_json::to_string(structured) {
                        Ok(json_string) => vec![Content::text(json_string)],
                        Err(e) => {
                            // Return error if we can't serialize the structured content
                            let error = crate::error::ToolCallError::Execution(
                                crate::error::ToolCallExecutionError::ResponseParsingError {
                                    reason: format!("Failed to serialize structured content: {e}"),
                                    raw_response: None,
                                },
                            );
                            return Err(error);
                        }
                    }
                } else {
                    vec![Content::text(response.to_mcp_content())]
                };

                // Return successful response
                Ok(CallToolResult {
                    content,
                    structured_content,
                    is_error: Some(!response.is_success),
                })
            }
            Err(e) => {
                // Return ToolCallError directly
                Err(e)
            }
        }
    }

    /// Execute tool and return raw HTTP response
    pub async fn execute(
        &self,
        arguments: &Value,
        authorization: Authorization,
    ) -> Result<crate::http_client::HttpResponse, crate::error::ToolCallError> {
        // Extract authorization header if present
        let auth_header: Option<&rmcp_actix_web::transport::AuthorizationHeader> =
            match &authorization {
                Authorization::None => None,
                #[cfg(feature = "authorization-token-passthrough")]
                Authorization::PassthroughWarn(header)
                | Authorization::PassthroughSilent(header) => header.as_ref(),
            };

        // Create HTTP client with authorization if provided
        let client = if let Some(auth) = auth_header {
            self.http_client.with_authorization(&auth.0)
        } else {
            self.http_client.clone()
        };

        // Execute the HTTP request using the (potentially auth-enhanced) HTTP client
        // Return the raw HttpResponse without MCP formatting
        client.execute_tool_call(&self.metadata, arguments).await
    }
}

/// MCP compliance - Convert Tool to rmcp::model::Tool
impl From<&Tool> for McpTool {
    fn from(tool: &Tool) -> Self {
        (&tool.metadata).into()
    }
}
