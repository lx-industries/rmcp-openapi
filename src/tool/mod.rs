pub mod metadata;
pub mod openapi_tool;

pub use metadata::ToolMetadata;
pub use openapi_tool::OpenApiTool;
