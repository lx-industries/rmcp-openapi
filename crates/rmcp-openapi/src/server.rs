use rmcp::{
    RoleServer, ServerHandler,
    model::{
        CallToolRequestParam, CallToolResult, ErrorData, Implementation, InitializeResult,
        ListToolsResult, PaginatedRequestParam, ProtocolVersion, ServerCapabilities,
        ToolsCapability,
    },
    service::RequestContext,
};
use serde_json::Value;

use reqwest::header::HeaderMap;
use url::Url;

use crate::error::Error;
use crate::tool::{Tool, ToolCollection, ToolMetadata};
use tracing::{debug, info, info_span, warn};

#[derive(Clone)]
pub struct Server {
    pub openapi_spec: serde_json::Value,
    pub tool_collection: ToolCollection,
    pub base_url: Url,
    pub default_headers: Option<HeaderMap>,
    pub tag_filter: Option<Vec<String>>,
    pub method_filter: Option<Vec<reqwest::Method>>,
}

impl Server {
    /// Create a new Server instance with required parameters
    pub fn new(
        openapi_spec: serde_json::Value,
        base_url: Url,
        default_headers: Option<HeaderMap>,
        tag_filter: Option<Vec<String>>,
        method_filter: Option<Vec<reqwest::Method>>,
    ) -> Self {
        Self {
            openapi_spec,
            tool_collection: ToolCollection::new(),
            base_url,
            default_headers,
            tag_filter,
            method_filter,
        }
    }

    /// Parse the `OpenAPI` specification and convert to OpenApiTool instances
    ///
    /// # Errors
    ///
    /// Returns an error if the spec cannot be parsed or tools cannot be generated
    pub fn load_openapi_spec(&mut self) -> Result<(), Error> {
        let span = info_span!("tool_registration");
        let _enter = span.enter();

        // Parse the OpenAPI specification
        let spec = crate::spec::Spec::from_value(self.openapi_spec.clone())?;

        // Generate OpenApiTool instances directly
        let tools = spec.to_openapi_tools(
            self.tag_filter.as_deref(),
            self.method_filter.as_deref(),
            Some(self.base_url.clone()),
            self.default_headers.clone(),
        )?;

        self.tool_collection = ToolCollection::from_tools(tools);

        info!(
            tool_count = self.tool_collection.len(),
            "Loaded tools from OpenAPI spec"
        );

        Ok(())
    }

    /// Get the number of loaded tools
    #[must_use]
    pub fn tool_count(&self) -> usize {
        self.tool_collection.len()
    }

    /// Get all tool names
    #[must_use]
    pub fn get_tool_names(&self) -> Vec<String> {
        self.tool_collection.get_tool_names()
    }

    /// Check if a specific tool exists
    #[must_use]
    pub fn has_tool(&self, name: &str) -> bool {
        self.tool_collection.has_tool(name)
    }

    /// Get a tool by name
    #[must_use]
    pub fn get_tool(&self, name: &str) -> Option<&Tool> {
        self.tool_collection.get_tool(name)
    }

    /// Get tool metadata by name
    #[must_use]
    pub fn get_tool_metadata(&self, name: &str) -> Option<&ToolMetadata> {
        self.get_tool(name).map(|tool| &tool.metadata)
    }

    /// Get basic tool statistics
    #[must_use]
    pub fn get_tool_stats(&self) -> String {
        self.tool_collection.get_stats()
    }

    /// Simple validation - check that tools are loaded
    ///
    /// # Errors
    ///
    /// Returns an error if no tools are loaded
    pub fn validate_registry(&self) -> Result<(), Error> {
        if self.tool_collection.is_empty() {
            return Err(Error::McpError("No tools loaded".to_string()));
        }
        Ok(())
    }
}

impl ServerHandler for Server {
    fn get_info(&self) -> InitializeResult {
        InitializeResult {
            protocol_version: ProtocolVersion::V_2024_11_05,
            server_info: Implementation {
                name: "OpenAPI MCP Server".to_string(),
                version: "0.1.0".to_string(),
            },
            capabilities: ServerCapabilities {
                tools: Some(ToolsCapability {
                    list_changed: Some(false),
                }),
                ..Default::default()
            },
            instructions: Some("Exposes OpenAPI endpoints as MCP tools".to_string()),
        }
    }

    async fn list_tools(
        &self,
        _request: Option<PaginatedRequestParam>,
        _context: RequestContext<RoleServer>,
    ) -> Result<ListToolsResult, ErrorData> {
        let span = info_span!("list_tools", tool_count = self.tool_collection.len());
        let _enter = span.enter();

        debug!("Processing MCP list_tools request");

        // Delegate to tool collection for MCP tool conversion
        let tools = self.tool_collection.to_mcp_tools();

        info!(
            returned_tools = tools.len(),
            "MCP list_tools request completed successfully"
        );

        Ok(ListToolsResult {
            tools,
            next_cursor: None,
        })
    }

    async fn call_tool(
        &self,
        request: CallToolRequestParam,
        _context: RequestContext<RoleServer>,
    ) -> Result<CallToolResult, ErrorData> {
        let span = info_span!(
            "call_tool",
            tool_name = %request.name
        );
        let _enter = span.enter();

        debug!(
            tool_name = %request.name,
            has_arguments = !request.arguments.as_ref().unwrap_or(&serde_json::Map::new()).is_empty(),
            "Processing MCP call_tool request"
        );

        let arguments = request.arguments.unwrap_or_default();
        let arguments_value = Value::Object(arguments);

        // Delegate all tool validation and execution to the tool collection
        match self
            .tool_collection
            .call_tool(&request.name, &arguments_value)
            .await
        {
            Ok(result) => {
                info!(
                    tool_name = %request.name,
                    success = true,
                    "MCP call_tool request completed successfully"
                );
                Ok(result)
            }
            Err(e) => {
                warn!(
                    tool_name = %request.name,
                    success = false,
                    error = %e,
                    "MCP call_tool request failed"
                );
                // Convert ToolCallError to ErrorData and return as error
                Err(e.into())
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::error::ToolCallValidationError;
    use crate::{ToolCallError, ToolMetadata};
    use serde_json::json;

    #[test]
    fn test_tool_not_found_error_with_suggestions() {
        // Create test tool metadata
        let tool1_metadata = ToolMetadata {
            name: "getPetById".to_string(),
            title: Some("Get Pet by ID".to_string()),
            description: "Find pet by ID".to_string(),
            parameters: json!({
                "type": "object",
                "properties": {
                    "petId": {
                        "type": "integer"
                    }
                },
                "required": ["petId"]
            }),
            output_schema: None,
            method: "GET".to_string(),
            path: "/pet/{petId}".to_string(),
        };

        let tool2_metadata = ToolMetadata {
            name: "getPetsByStatus".to_string(),
            title: Some("Find Pets by Status".to_string()),
            description: "Find pets by status".to_string(),
            parameters: json!({
                "type": "object",
                "properties": {
                    "status": {
                        "type": "array",
                        "items": {
                            "type": "string"
                        }
                    }
                },
                "required": ["status"]
            }),
            output_schema: None,
            method: "GET".to_string(),
            path: "/pet/findByStatus".to_string(),
        };

        // Create OpenApiTool instances
        let tool1 = Tool::new(tool1_metadata, None, None).unwrap();
        let tool2 = Tool::new(tool2_metadata, None, None).unwrap();

        // Create server with tools
        let mut server = Server::new(
            serde_json::Value::Null,
            url::Url::parse("http://example.com").unwrap(),
            None,
            None,
            None,
        );
        server.tool_collection = ToolCollection::from_tools(vec![tool1, tool2]);

        // Test: Create ToolNotFound error with a typo
        let tool_names = server.get_tool_names();
        let tool_name_refs: Vec<&str> = tool_names.iter().map(|s| s.as_str()).collect();

        let error = ToolCallError::Validation(ToolCallValidationError::tool_not_found(
            "getPetByID".to_string(),
            &tool_name_refs,
        ));
        let error_data: ErrorData = error.into();
        let error_json = serde_json::to_value(&error_data).unwrap();

        // Snapshot the error to verify suggestions
        insta::assert_json_snapshot!(error_json);
    }

    #[test]
    fn test_tool_not_found_error_no_suggestions() {
        // Create test tool metadata
        let tool_metadata = ToolMetadata {
            name: "getPetById".to_string(),
            title: Some("Get Pet by ID".to_string()),
            description: "Find pet by ID".to_string(),
            parameters: json!({
                "type": "object",
                "properties": {
                    "petId": {
                        "type": "integer"
                    }
                },
                "required": ["petId"]
            }),
            output_schema: None,
            method: "GET".to_string(),
            path: "/pet/{petId}".to_string(),
        };

        // Create OpenApiTool instance
        let tool = Tool::new(tool_metadata, None, None).unwrap();

        // Create server with tool
        let mut server = Server::new(
            serde_json::Value::Null,
            url::Url::parse("http://example.com").unwrap(),
            None,
            None,
            None,
        );
        server.tool_collection = ToolCollection::from_tools(vec![tool]);

        // Test: Create ToolNotFound error with unrelated name
        let tool_names = server.get_tool_names();
        let tool_name_refs: Vec<&str> = tool_names.iter().map(|s| s.as_str()).collect();

        let error = ToolCallError::Validation(ToolCallValidationError::tool_not_found(
            "completelyUnrelatedToolName".to_string(),
            &tool_name_refs,
        ));
        let error_data: ErrorData = error.into();
        let error_json = serde_json::to_value(&error_data).unwrap();

        // Snapshot the error to verify no suggestions
        insta::assert_json_snapshot!(error_json);
    }

    #[test]
    fn test_validation_error_converted_to_error_data() {
        // Test that validation errors are properly converted to ErrorData
        let error = ToolCallError::Validation(ToolCallValidationError::InvalidParameters {
            violations: vec![crate::error::ValidationError::invalid_parameter(
                "page".to_string(),
                &["page_number".to_string(), "page_size".to_string()],
            )],
        });

        let error_data: ErrorData = error.into();
        let error_json = serde_json::to_value(&error_data).unwrap();

        // Verify the basic structure
        assert_eq!(error_json["code"], -32602); // Invalid params error code

        // Snapshot the full error to verify the new error message format
        insta::assert_json_snapshot!(error_json);
    }
}
