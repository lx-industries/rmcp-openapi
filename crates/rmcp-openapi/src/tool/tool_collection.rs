use super::Tool;
use crate::config::Authorization;
use crate::error::{ToolCallError, ToolCallValidationError};
use rmcp::model::{CallToolResult, Tool as McpTool};
use serde_json::Value;
use tracing::debug_span;

/// Collection of tools with built-in validation and lookup capabilities
///
/// This struct encapsulates all tool management logic in the library layer,
/// providing a clean API for the binary to delegate tool operations to.
#[derive(Clone, Default)]
pub struct ToolCollection {
    tools: Vec<Tool>,
}

impl ToolCollection {
    /// Create a new empty tool collection
    pub fn new() -> Self {
        Self { tools: Vec::new() }
    }

    /// Create a tool collection from a vector of tools
    pub fn from_tools(tools: Vec<Tool>) -> Self {
        Self { tools }
    }

    /// Add a tool to the collection
    pub fn add_tool(&mut self, tool: Tool) {
        self.tools.push(tool);
    }

    /// Get the number of tools in the collection
    pub fn len(&self) -> usize {
        self.tools.len()
    }

    /// Check if the collection is empty
    pub fn is_empty(&self) -> bool {
        self.tools.is_empty()
    }

    /// Get all tool names
    pub fn get_tool_names(&self) -> Vec<String> {
        self.tools
            .iter()
            .map(|tool| tool.metadata.name.clone())
            .collect()
    }

    /// Check if a specific tool exists
    pub fn has_tool(&self, name: &str) -> bool {
        self.tools.iter().any(|tool| tool.metadata.name == name)
    }

    /// Get a tool by name
    pub fn get_tool(&self, name: &str) -> Option<&Tool> {
        self.tools.iter().find(|tool| tool.metadata.name == name)
    }

    /// Convert all tools to MCP Tool format for list_tools response
    pub fn to_mcp_tools(&self) -> Vec<McpTool> {
        self.tools.iter().map(McpTool::from).collect()
    }

    /// Call a tool by name with validation
    ///
    /// This method encapsulates all tool validation logic:
    /// - Tool not found errors with suggestions
    /// - Parameter validation
    /// - Tool execution
    pub async fn call_tool(
        &self,
        tool_name: &str,
        arguments: &Value,
        authorization: Authorization,
    ) -> Result<CallToolResult, ToolCallError> {
        let span = debug_span!(
            "tool_execution",
            tool_name = %tool_name,
            total_tools = self.tools.len()
        );
        let _enter = span.enter();

        // First validate that the tool exists
        if let Some(tool) = self.get_tool(tool_name) {
            // Tool exists, delegate to the tool's call method
            tool.call(arguments, authorization).await
        } else {
            // Tool not found - generate suggestions and return validation error
            let tool_names: Vec<&str> = self
                .tools
                .iter()
                .map(|tool| tool.metadata.name.as_str())
                .collect();

            Err(ToolCallError::Validation(
                ToolCallValidationError::tool_not_found(tool_name.to_string(), &tool_names),
            ))
        }
    }

    /// Get basic statistics about the tool collection
    pub fn get_stats(&self) -> String {
        format!("Total tools: {}", self.tools.len())
    }

    /// Get an iterator over the tools
    pub fn iter(&self) -> impl Iterator<Item = &Tool> {
        self.tools.iter()
    }
}

impl From<Vec<Tool>> for ToolCollection {
    fn from(tools: Vec<Tool>) -> Self {
        Self::from_tools(tools)
    }
}

impl IntoIterator for ToolCollection {
    type Item = Tool;
    type IntoIter = std::vec::IntoIter<Tool>;

    fn into_iter(self) -> Self::IntoIter {
        self.tools.into_iter()
    }
}

impl<'a> IntoIterator for &'a ToolCollection {
    type Item = &'a Tool;
    type IntoIter = std::slice::Iter<'a, Tool>;

    fn into_iter(self) -> Self::IntoIter {
        self.tools.iter()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tool::ToolMetadata;
    use serde_json::json;

    fn create_test_tool(name: &str, description: &str) -> Tool {
        let metadata = ToolMetadata {
            name: name.to_string(),
            title: Some(name.to_string()),
            description: Some(description.to_string()),
            parameters: json!({
                "type": "object",
                "properties": {
                    "id": {"type": "integer"}
                },
                "required": ["id"]
            }),
            output_schema: None,
            method: "GET".to_string(),
            path: format!("/{}", name),
            security: None,
            parameter_mappings: std::collections::HashMap::new(),
        };
        Tool::new(metadata, None, None).unwrap()
    }

    #[test]
    fn test_tool_collection_creation() {
        let collection = ToolCollection::new();
        assert_eq!(collection.len(), 0);
        assert!(collection.is_empty());
    }

    #[test]
    fn test_tool_collection_from_tools() {
        let tool1 = create_test_tool("test1", "Test tool 1");
        let tool2 = create_test_tool("test2", "Test tool 2");
        let tools = vec![tool1, tool2];

        let collection = ToolCollection::from_tools(tools);
        assert_eq!(collection.len(), 2);
        assert!(!collection.is_empty());
        assert!(collection.has_tool("test1"));
        assert!(collection.has_tool("test2"));
        assert!(!collection.has_tool("test3"));
    }

    #[test]
    fn test_add_tool() {
        let mut collection = ToolCollection::new();
        let tool = create_test_tool("test", "Test tool");

        collection.add_tool(tool);
        assert_eq!(collection.len(), 1);
        assert!(collection.has_tool("test"));
    }

    #[test]
    fn test_get_tool_names() {
        let tool1 = create_test_tool("getPetById", "Get pet by ID");
        let tool2 = create_test_tool("getPetsByStatus", "Get pets by status");
        let collection = ToolCollection::from_tools(vec![tool1, tool2]);

        let names = collection.get_tool_names();
        assert_eq!(names, vec!["getPetById", "getPetsByStatus"]);
    }

    #[test]
    fn test_get_tool() {
        let tool = create_test_tool("test", "Test tool");
        let collection = ToolCollection::from_tools(vec![tool]);

        assert!(collection.get_tool("test").is_some());
        assert!(collection.get_tool("nonexistent").is_none());
    }

    #[test]
    fn test_to_mcp_tools() {
        let tool1 = create_test_tool("test1", "Test tool 1");
        let tool2 = create_test_tool("test2", "Test tool 2");
        let collection = ToolCollection::from_tools(vec![tool1, tool2]);

        let mcp_tools = collection.to_mcp_tools();
        assert_eq!(mcp_tools.len(), 2);
        assert_eq!(mcp_tools[0].name, "test1");
        assert_eq!(mcp_tools[1].name, "test2");
    }

    #[actix_web::test]
    async fn test_call_tool_not_found_with_suggestions() {
        let tool1 = create_test_tool("getPetById", "Get pet by ID");
        let tool2 = create_test_tool("getPetsByStatus", "Get pets by status");
        let collection = ToolCollection::from_tools(vec![tool1, tool2]);

        let result = collection
            .call_tool("getPetByID", &json!({}), Authorization::default())
            .await;
        assert!(result.is_err());

        if let Err(ToolCallError::Validation(ToolCallValidationError::ToolNotFound {
            tool_name,
            suggestions,
        })) = result
        {
            assert_eq!(tool_name, "getPetByID");
            // The algorithm finds multiple similar matches
            assert!(suggestions.contains(&"getPetById".to_string()));
            assert!(!suggestions.is_empty());
        } else {
            panic!("Expected ToolNotFound error with suggestions");
        }
    }

    #[actix_web::test]
    async fn test_call_tool_not_found_no_suggestions() {
        let tool = create_test_tool("getPetById", "Get pet by ID");
        let collection = ToolCollection::from_tools(vec![tool]);

        let result = collection
            .call_tool(
                "completelyDifferentName",
                &json!({}),
                Authorization::default(),
            )
            .await;
        assert!(result.is_err());

        if let Err(ToolCallError::Validation(ToolCallValidationError::ToolNotFound {
            tool_name,
            suggestions,
        })) = result
        {
            assert_eq!(tool_name, "completelyDifferentName");
            assert!(suggestions.is_empty());
        } else {
            panic!("Expected ToolNotFound error with no suggestions");
        }
    }

    #[test]
    fn test_iterators() {
        let tool1 = create_test_tool("test1", "Test tool 1");
        let tool2 = create_test_tool("test2", "Test tool 2");
        let collection = ToolCollection::from_tools(vec![tool1, tool2]);

        // Test iter()
        let names: Vec<String> = collection
            .iter()
            .map(|tool| tool.metadata.name.clone())
            .collect();
        assert_eq!(names, vec!["test1", "test2"]);

        // Test IntoIterator for &collection
        let names: Vec<String> = (&collection)
            .into_iter()
            .map(|tool| tool.metadata.name.clone())
            .collect();
        assert_eq!(names, vec!["test1", "test2"]);

        // Test IntoIterator for collection (consumes it)
        let names: Vec<String> = collection
            .into_iter()
            .map(|tool| tool.metadata.name.clone())
            .collect();
        assert_eq!(names, vec!["test1", "test2"]);
    }

    #[test]
    fn test_from_vec() {
        let tool1 = create_test_tool("test1", "Test tool 1");
        let tool2 = create_test_tool("test2", "Test tool 2");
        let tools = vec![tool1, tool2];

        let collection: ToolCollection = tools.into();
        assert_eq!(collection.len(), 2);
    }
}
