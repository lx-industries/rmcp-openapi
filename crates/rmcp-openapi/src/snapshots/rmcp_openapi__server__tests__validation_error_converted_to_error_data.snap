---
source: crates/rmcp-openapi/src/server.rs
expression: error_json
---
{
  "code": -32602,
  "message": "Validation failed - invalid parameter 'page' (suggestions: page_size, page_number)",
  "data": {
    "type": "validation-errors",
    "violations": [
      {
        "type": "invalid-parameter",
        "parameter": "page",
        "suggestions": [
          "page_size",
          "page_number"
        ],
        "valid_parameters": [
          "page_number",
          "page_size"
        ]
      }
    ]
  }
}
